### ISSUE TITLE
[HOLD App/issues/16660] [$2000] Pinch to zoom on a Macbook browser dynamically changes the app's view and components

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#17246]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. When a chat is open, pinch to zoom on either the LHN or the details tab.
 1 (alternate). On a fresh sign in page, zoom in on any part of the site with pinch to zoom.
## Expected Result
The page doesn't reorganize or shfit as you zoom in this way, this should be equivalent to magnification and shouldn't change the content ordering or structure of the app.
## Actual Result
App re-renders into different various forms as you zoom in breaking your ability to zoom into

## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.98-1
**Reproducible in staging?:** needs reproduction
**Reproducible in production?:** needs reproduction
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/231004704-3ffa4eb5-bcb9-456f-83c7-555e63e266b5.mov

**Expensify/Expensify Issue URL:**
**Issue reported by:** @johnmlee101 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681152993219619

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01caf96ff9d8613c10</li>
        <li>Upwork Job ID: 1646608088499531776</li>
        <li>Last Price Increase: 2023-05-02</li>
    </ul>
</details>

**Comment by alexpensify** [comment]
I'll try to test soon.

**Comment by alexpensify** [comment]
I tested here is my page before zoom 

![image](https://user-images.githubusercontent.com/8107070/231872864-8036ecbd-30eb-4a61-9b69-7b4453039b53.png)

Here is my page when zooming

![image](https://user-images.githubusercontent.com/8107070/231872911-b93a38a6-3822-4124-88fb-8529385e9739.png)

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01caf96ff9d8613c10

**Comment by MelvinBot** [comment]
Current assignee @alexpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @fedirjh (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @tgolen (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by Armedin** [comment]
This issue seems to happen whenever you zoom in on a web browser. The `dimensionsEventListener` is registering zoom events as a window dimension change. For instance, if I zoom 3 times and print the window dimension:

<img width="543" alt="Screenshot at Apr 14 00-26-02" src="https://user-images.githubusercontent.com/24688234/231904113-a926615f-2731-4f4a-ba5b-6a36754f76ef.png">

**Comment by alitoshmatov** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
 Pinch to zoom on a Macbook browser dynamically changes the app's view and components
### What is the root cause of that problem?
This is caused by how `react-native-web` handling dimensions.
https://github.com/Expensify/react-native-web/blob/bb2f765614628eeccdd9d2850cb287557a92b6c9/packages/react-native-web/src/exports/Dimensions/index.js#L63-L71
Above you can see it is setting width and height based on `window.visualViewport` which changes when you zoom. 
### What changes do you think we should make in order to solve the problem?
There is an option to make an upstream fix in `react-native-web` to set window height and width to `documentElement.clientHeight` which won't change when you zoom and feels more natural to set this measures to window object in dimensions. 

### What alternative solutions did you explore? (Optional)

**Comment by alexpensify** [comment]
@fedirjh - can you share feedback on this new proposal? Thanks!

**Comment by tgolen** [comment]
Thanks for the research into this issue. One thing that isn't clear to me is _why_ this is the expected behavior. Can anyone add some details about that? Is it because that's what happens on normal websites when you pinch to zoom?

**Comment by tgolen** [comment]
Great, thank you! That is an excellent explanation. Sounds like the next step for this is still waiting on an answer for https://github.com/Expensify/App/issues/17246#issuecomment-1510892091

**Comment by alitoshmatov** [comment]
@fedirjh I am suggestion to pass `visualViewport` width and height seperately along window, and screen values:
https://github.com/Expensify/react-native-web/blob/ccfd936f274ca2105745f9edbbb4aad80725e181/packages/react-native-web/src/exports/Dimensions/index.js#L63-L86

And then we can modify here to use `visualViewport` width:
https://github.com/rawalyogendra/App/blob/6a1ae162886703a772d27081eff4c99a49dbf971/src/components/ScreenWrapper/index.js#L92

But there is a concern which this solution might break philosophy of platform independent code. Since this code `visualViewport` will only be available in web platforms through `react-native-web`.

**Comment by alitoshmatov** [comment]
@fedirjh Yes in previous comment i meant window=documentElement, screen, and visualViewport 3 seperate fields.


> What if we pass the height of the visualViewport and the width of the documentElement ? Would that address both issues ?

Well it seems like there are some components which is styled based on `window.height` and they are adopting to smaller screen when zoomed, which is not desirable.

**Comment by alitoshmatov** [comment]
You are right, but I think https://github.com/Expensify/react-native-web/commit/ccfd936f274ca2105745f9edbbb4aad80725e181 itself is a workaround solution in expense of this issue. We should really reconsider this solution and https://github.com/Expensify/App/pull/14392 .

**Comment by tgolen** [comment]
I agree with you both. For this issue, let's make sure we revert the previous workaround and have a solution which fixes both problems 👍

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by alexpensify** [comment]
@fedirjh - are we good with the proposal now and can we carry on to the next steps? Thanks!

**Comment by alexpensify** [comment]
Whoops, I jumped the start line here. Thank you for clarifying!

**Comment by tgolen** [comment]
No new proposals yet

**Comment by tgolen** [comment]
Still no new proposals.

**Comment by tgolen** [comment]
Let's just let the price double on this next week and hopefully we'll start getting some proposals then.

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by tgolen** [comment]
Funny you should mention that, Melvin. Looks like great minds think alike and I beat you to the punch.

**Comment by tgolen** [comment]
Another day another similar update. Have a good weekend! 👋

**Comment by alexpensify** [comment]
Sigh, Melvin is still going to bark at us over the weekend.

**Comment by tgolen** [comment]
Cool, is the price increasing on this one today?

**Comment by tgolen** [comment]
@alexpensify any idea why the price hasn't increased for this?

**Comment by alexpensify** [comment]
No, I'm not sure why.

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $2000

**Comment by alexpensify** [comment]
The price has been updated and we are still taking proposals here.

**Comment by tgolen** [comment]
Still waiting

**Comment by alitoshmatov** [comment]
@fedirjh I think I have found potential solution: 
So it looks like we have [scale](https://developer.mozilla.org/en-US/docs/Web/API/VisualViewport/scale) field in visualViewport and it represents pinch zoom.

I think we should multiply it with height and width here:
https://github.com/Expensify/react-native-web/blob/ccfd936f274ca2105745f9edbbb4aad80725e181/packages/react-native-web/src/exports/Dimensions/index.js#L65-L66

And It results in correct representation of window measures, even if we are pinch zoomed

**Comment by alitoshmatov** [comment]
It is working as expected: 

https://user-images.githubusercontent.com/59907218/236642314-cd540f28-9147-45e4-a292-ff900b3a15bb.mov

**Comment by alitoshmatov** [comment]
@fedirjh No, It will be in react-native-web.

**Comment by alitoshmatov** [comment]
Before https://github.com/Expensify/App/pull/14392 `window` object provided by `Dimensions` didn't change in pinch zoom, so it was it is supposed behavior. As you said earlier, we are going to fix https://github.com/Expensify/App/pull/14392 so that it will react to ios keyboard openings, but won't change default behavior of `Dimensions.window`

**Comment by alitoshmatov** [comment]
No I mean, we do this:
```javascript
    height = Math.round(visualViewport.height * visualViewport.scale);
    width = Math.round(visualViewport.width * visualViewport.scale);
``` 
in here https://github.com/Expensify/react-native-web/blob/ccfd936f274ca2105745f9edbbb4aad80725e181/packages/react-native-web/src/exports/Dimensions/index.js#L65-L66

We do not touch the `Dimensions.window.scale`

**Comment by alitoshmatov** [comment]
I will get back to you with testing results, tomorrow

**Comment by alitoshmatov** [comment]
@fedirjh So it looks like, pinch zooming is prevented on mWeb(both: android and ios). But I couldn't locate exact code responsible for this. But I think pinch zoom not working in mWebs is not in capacity of this issue. 

I have tested if opening keyboard is still behaving as expected, and it does:
https://user-images.githubusercontent.com/59907218/236671233-060cc9d6-f651-4809-9456-51184a31e6f0.mp4

**Comment by tgolen** [comment]
It sounds like the proposal will be a good incremental fix, but for our App, we push for full platform support across mWeb (mobile chrome and mobile safari). Ideally, we would be able to figure out why the pinch zooming is prevented.

**Comment by alitoshmatov** [comment]
@tgolen I think pinch zoom not working in mobile should be separate issue and treated as one, since it has neither the same root cause nor the same solution as current issue.

**Comment by alitoshmatov** [comment]
@fedirjh It is an existing issue, where safari tabs minimizes app content won't stretch to occupy. It is still present without any changes
<img width="360" alt="Screenshot 2023-05-09 at 11 08 37 AM" src="https://user-images.githubusercontent.com/59907218/237008737-c16fc30e-af1d-41b4-911d-67800f53678a.png">
<img width="344" alt="Screenshot 2023-05-09 at 11 09 20 AM" src="https://user-images.githubusercontent.com/59907218/237008871-376f3603-1804-42c4-9bac-e3228f7d6ebf.png">

**Comment by alitoshmatov** [comment]
@fedirjh Can you clarify then all objectives of this issue here. I am confused, as of which issues are present. Also can we clarify that  if mobile pinch zoom is not intentionally disabled and should really be fixed

**Comment by alitoshmatov** [comment]
Pinch zoom is not working when `FlatList` is inverted, if I remove this, we can zoom in chat
https://github.com/Expensify/App/blob/6176fcf28769f64b5155f019a4216575a4b87eaa/src/components/InvertedFlatList/index.js#LL44C10-L44C10

This is some results where it lead me:
https://github.com/Expensify/react-native-web/blob/e68c32770757194af103cca0095c0c204995505b/packages/react-native-web/src/vendor/react-native/FlatList/index.js#L117-L120
https://github.com/Expensify/react-native-web/blob/e202ba84bbf6ee296327eec2d5d8165fbaaf19e0/packages/react-native-web/src/vendor/react-native/VirtualizedList/index.js#L2349-L2351

But, with just `transform: scale(-1)` I couldn't recreate it, where pinch zoom stop working.

**Comment by tgolen** [comment]
OK, that sounds fine to me. Thank you! @alexpensify 🟢 to hire @alitoshmatov for this.

**Comment by alexpensify** [comment]
Hmmm, I assigned @alitoshmatov but waiting for the automation to kick in here. Melvin?

**Comment by alexpensify** [comment]
There we go, good catch @fedirjh!

**Comment by alitoshmatov** [comment]
Applied to the upwork job, the pr will be ready in couple of hours.

**Comment by alitoshmatov** [comment]
@fedirjh
Opened PR in react-native-web -  https://github.com/Expensify/react-native-web/pull/18/files

I am not sure how deployment works in this package, looks like it is deployed to npm automatically. I open PR in expensify app with bumping the package version once aboce pr is merged

**Comment by alitoshmatov** [comment]
@fedirjh should I also keep this https://github.com/Expensify/react-native-web/pull/18/files or remove it?

**Comment by alitoshmatov** [comment]
@fedirjh Looks like expsify/react-native-web and  https://github.com/necolas/react-native-web doesn't much and I am getting a lot of changes when trying to open PR with my current branch 
<img width="1373" alt="Screenshot 2023-05-10 at 11 21 36 PM" src="https://github.com/Expensify/App/assets/59907218/8abd6d3f-2d92-48ab-af68-2e858d633a51">

Should I open new branch in upstream and then open a pr?

**Comment by alexpensify** [comment]
Easy Melvin, the PR has been merged.

**Comment by alitoshmatov** [comment]
@fedirjh Finally opened PR in upstream - https://github.com/necolas/react-native-web/pull/2520

**Comment by tgolen** [comment]
Dropping to weekly priority now that a PR is being worked on.

**Comment by strepanier03** [comment]
👋 hey, quick note that we should keep all e/app GHs as daily unless they have "Hold" in the title. I'm going to flip this back to daily as I'm doing the resync and it's [step in the SO process guide](https://stackoverflowteams.com/c/expensify/questions/8208/14658#14658).

**Comment by alitoshmatov** [comment]
@tgolen Can you check this out, and help us to deploy `@expensify/react-native-web`

**Comment by tgolen** [comment]
That package is deployed now. What are the next steps?

**Comment by alitoshmatov** [comment]
@tgolen We were having problems related to breaking changes introduced in https://github.com/Expensify/react-native-web/pull/17

https://expensify.slack.com/archives/C01GTK53T8Q/p1684256093521369?thread_ts=1684080150.427299&cid=C01GTK53T8Q

**Comment by alexpensify** [comment]
@dylanexpensify - this one is moving forward, but assigning if a BZ team member is needed to step in while I'm OOO.

**Comment by dylanexpensify** [comment]
Nice! Let's see what @alitoshmatov comes up with here.

**Comment by dylanexpensify** [comment]
[I'm heading OOO tomorrow](https://expensify.slack.com/archives/C01SKUP7QR0/p1684776773511559) for about a week so adding another BZ member while I'm away to help keep the train moving! 🚂 Thanks @NicMendonca!!

**Comment by NicMendonca** [comment]
Going OOO until June 5th so assigning a buddy to this GH to watch over the issue. I'll pick it back up when I am back from OOO if this is still open.

**Comment by NicMendonca** [comment]
@alexpensify actually, I will keep you assigned since you are also on Bug0 team 😄

**Comment by tgolen** [comment]
Bumping [the slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1684832214218819?thread_ts=1684080150.427299&cid=C01GTK53T8Q) to see what the next steps are.

**Comment by tgolen** [comment]
OK, sounds good to me. I'll place this on HOLD for now then. Thanks!

**Comment by alexpensify** [comment]
@dylanexpensify and @NicMendonca - I appreciate your help here! I'm back online and removing your assignment.

**Comment by alexpensify** [comment]
@tgolen -  with Rory's feedback here https://github.com/Expensify/App/issues/16660#issuecomment-1560003585, should we make this one a Weekly now?

**Comment by alexpensify** [comment]
Thank you, Tim!

**Comment by alexpensify** [comment]
@fedirjh - when you get a chance, can we get feedback on the next steps here and the current proposals? Thanks!

**Comment by alexpensify** [comment]
@fedirjh - disregard, I mixed this one up with another GH. For now, I'm removing my last comment to avoid more confusion.

**Comment by alexpensify** [comment]
Still on hold here

**Comment by alexpensify** [comment]
Still on hold

**Comment by alexpensify** [comment]
On hold

**Comment by alexpensify** [comment]
Still on hold here

**Comment by alexpensify** [comment]
On hold here, #1600 is still accepting proposals.

**Comment by alexpensify** [comment]
Still on hold here

**Comment by alexpensify** [comment]
On hold

**Comment by alexpensify** [comment]
Still on hold here

**Comment by alexpensify** [comment]
Weekly update: On hold

**Comment by alexpensify** [comment]
Weekly update: On hold

**Comment by alexpensify** [comment]
Weekly update: On hold

**Comment by alexpensify** [comment]
Weekly update: On hold

**Comment by alexpensify** [comment]
Weekly update: On hold but looks like there is movement in #1600

**Comment by johnmlee101** [comment]
Hmm, that goes to an unrelated PR?

**Comment by alexpensify** [comment]
Whoops, thanks for flagging, https://github.com/Expensify/App/issues/16660 should be the correct one

**Comment by alexpensify** [comment]
Weekly update: On Hold

**Comment by alexpensify** [comment]
Weekly update: On Hold

**Comment by alexpensify** [comment]
Weekly update: On Hold

**Comment by alexpensify** [comment]
Weekly update: On Hold

**Comment by alexpensify** [comment]
Weekly update: On Hold

**Comment by alexpensify** [comment]
Weekly update: On Hold but looks like the PR might get moving again

**Comment by alexpensify** [comment]
Weekly Update: There should be a PR soon for the GH that put this one on hold

**Comment by alexpensify** [comment]
Weekly update: On Hold

**Comment by alexpensify** [comment]
Weekly update: On Hold but we had some GH movement yesterday. 🤞🏼

**Comment by alexpensify** [comment]
Weekly update: On Hold

**Comment by alexpensify** [comment]
Thank you for that update @fedirjh!

**Comment by alexpensify** [comment]
@fedirjh - I realized that I don't think the PR you mentioned last week is in this GH. Can you please share the link since the one in the title has many PRs associated with it now? Thanks!

**Comment by alexpensify** [comment]
It is time! @tgolen and @fedirjh are you OK for me to take this one off hold and move it back to daily? I'm assuming that we will need updated proposals but keep me posted on the next steps. We can prepare accordingly. Thanks!

**Comment by alitoshmatov** [comment]
https://github.com/Expensify/App/pull/24482 is live, which means we can take this issue off hold, finally 🥳

**Comment by alexpensify** [comment]
Ok, I chatted with @tgolen. Let's keep this one on hold to clear the regression period. 

@fedirjh - Sorry, I'm confused are you suggesting that the PR (https://github.com/Expensify/App/pull/24482) for https://github.com/Expensify/App/issues/16660 fixed this issue too? If yes, then we will need to prepare for partial payment here instead of another fix when the HOLD is removed in this GitHub? I want to confirm the required actions since I will be OOO next week. If there is no new fix for this issue, then can find a replacement to take over the payment process in Upwork. Thanks!

**Comment by alexpensify** [comment]
Got it, we will close after the regression period and prepare for the payment period. Thanks!

**Comment by alexpensify** [comment]
Reassigning another team member, I'm going OOO until Monday, November 27, and will take it back if it's still open by my return date.

@laurenreidexpensify - Required action from the team:

There will need to be a payment after the regression period closes while I'm offline next week. I haven't been able to write up a payment summary but will need help with that one before the payment date and then pay everyone via Upwork. Thanks!

**Comment by laurenreidexpensify** [comment]
Ready for payment, regression period has passed as https://github.com/Expensify/App/pull/24482 has been on staging for a week

**Comment by laurenreidexpensify** [comment]
Payment Summary: 

- Contributor @alitoshmatov - $2000, payment will be in Upwork
- C+ @fedirjh - $2000, payment will be in Upwork

**Comment by laurenreidexpensify** [comment]
Offers sent in Upwork, 

@fedirjh is there a regression test that we need to add for this one?

**Comment by alitoshmatov** [comment]
Accepted the offer

**Comment by laurenreidexpensify** [comment]
Payment Summary:

- Contributor @alitoshmatov - $2000, payment issued in Upwork
- C+ @fedirjh - $2000, payment issued in Upwork

**Comment by laurenreidexpensify** [comment]
Outstanding action: waiting for @fedirjh to confirm if we need a regression test

**Comment by alexpensify** [comment]
Thanks for your help here @laurenreidexpensify to complete the payment process!
